File: backend/app/models/system.py
from sqlalchemy import Boolean, Column, ForeignKey, Integer, String, Table, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base
from app.models.base import TimestampMixin

user_roles = Table(
    "user_roles",
    Base.metadata,
    Column("user_id", ForeignKey("users.id", ondelete="CASCADE"), primary_key=True),
    Column("role_id", ForeignKey("roles.id", ondelete="CASCADE"), primary_key=True),
)

role_permissions = Table(
    "role_permissions",
    Base.metadata,
    Column("role_id", ForeignKey("roles.id", ondelete="CASCADE"), primary_key=True),
    Column("permission_id", ForeignKey("permissions.id", ondelete="CASCADE"), primary_key=True),
)

role_menus = Table(
    "role_menus",
    Base.metadata,
    Column("role_id", ForeignKey("roles.id", ondelete="CASCADE"), primary_key=True),
    Column("menu_id", ForeignKey("menus.id", ondelete="CASCADE"), primary_key=True),
)


class Department(TimestampMixin, Base):
    __tablename__ = "departments"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(80), unique=True, index=True)
    parent_id: Mapped[int | None] = mapped_column(ForeignKey("departments.id"))
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)


class User(TimestampMixin, Base):
    __tablename__ = "users"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    username: Mapped[str] = mapped_column(String(60), unique=True, index=True)
    password_hash: Mapped[str] = mapped_column(String(255))
    display_name: Mapped[str] = mapped_column(String(80))
    mobile: Mapped[str | None] = mapped_column(String(20), index=True)
    email: Mapped[str | None] = mapped_column(String(120))
    department_id: Mapped[int | None] = mapped_column(ForeignKey("departments.id"))
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    is_superuser: Mapped[bool] = mapped_column(Boolean, default=False)


class Role(TimestampMixin, Base):
    __tablename__ = "roles"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(80), unique=True, index=True)
    code: Mapped[str] = mapped_column(String(80), unique=True, index=True)
    description: Mapped[str | None] = mapped_column(String(255))
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)


class Permission(TimestampMixin, Base):
    __tablename__ = "permissions"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(String(80))
    code: Mapped[str] = mapped_column(String(120), unique=True, index=True)
    description: Mapped[str | None] = mapped_column(String(255))


class Menu(TimestampMixin, Base):
    __tablename__ = "menus"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    parent_id: Mapped[int | None] = mapped_column(ForeignKey("menus.id"))
    title: Mapped[str] = mapped_column(String(80))
    path: Mapped[str] = mapped_column(String(160), index=True)
    component: Mapped[str | None] = mapped_column(String(160))
    icon: Mapped[str | None] = mapped_column(String(80))
    permission_code: Mapped[str | None] = mapped_column(String(120))
    sort_order: Mapped[int] = mapped_column(Integer, default=0)
    is_visible: Mapped[bool] = mapped_column(Boolean, default=True)


class OperationLog(TimestampMixin, Base):
    __tablename__ = "operation_logs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"))
    action: Mapped[str] = mapped_column(String(120), index=True)
    resource: Mapped[str | None] = mapped_column(String(120))
    method: Mapped[str | None] = mapped_column(String(12))
    path: Mapped[str | None] = mapped_column(String(255))
    ip_address: Mapped[str | None] = mapped_column(String(64))
    detail: Mapped[str | None] = mapped_column(Text)


class LoginLog(TimestampMixin, Base):
    __tablename__ = "login_logs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    username: Mapped[str] = mapped_column(String(60), index=True)
    user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"))
    ip_address: Mapped[str | None] = mapped_column(String(64))
    user_agent: Mapped[str | None] = mapped_column(String(255))
    success: Mapped[bool] = mapped_column(Boolean, default=False)
    message: Mapped[str | None] = mapped_column(String(255))

File: frontend/src/api/system.ts
import { request } from './request'
import type { OperationLog, PageResponse, PermissionItem, SystemDepartment, SystemMenu, SystemRole, SystemUser } from '@/types'

export const systemApi = {
  listUsers(params?: { keyword?: string; page?: number; page_size?: number }) {
    return request.get<PageResponse<SystemUser>>('/system/users', { params })
  },
  createUser(data: {
    username: string
    password: string
    display_name: string
    mobile?: string | null
    email?: string | null
    is_active?: boolean
    is_superuser?: boolean
  }) {
    return request.post<SystemUser>('/system/users', data)
  },
  updateUser(userId: number, data: Partial<Pick<SystemUser, 'display_name' | 'mobile' | 'email' | 'is_active' | 'is_superuser'>>) {
    return request.patch<SystemUser>(`/system/users/${userId}`, data)
  },
  listRoles(params?: { page?: number; page_size?: number }) {
    return request.get<PageResponse<SystemRole>>('/system/roles', { params })
  },
  createRole(data: { name: string; code: string; description?: string | null; is_active?: boolean }) {
    return request.post<SystemRole>('/system/roles', data)
  },
  updateRole(roleId: number, data: Partial<Pick<SystemRole, 'name' | 'description' | 'is_active'>>) {
    return request.patch<SystemRole>(`/system/roles/${roleId}`, data)
  },
  listPermissions() {
    return request.get<PermissionItem[]>('/system/permissions')
  },
  getRolePermissions(roleId: number) {
    return request.get<string[]>(`/system/roles/${roleId}/permissions`)
  },
  updateRolePermissions(roleId: number, codes: string[]) {
    return request.put<string[]>(`/system/roles/${roleId}/permissions`, { codes })
  },
  listDepartments(params?: { page?: number; page_size?: number }) {
    return request.get<PageResponse<SystemDepartment>>('/system/departments', { params })
  },
  createDepartment(data: { name: string; parent_id?: number | null; sort_order?: number; is_active?: boolean }) {
    return request.post<SystemDepartment>('/system/departments', data)
  },
  updateDepartment(departmentId: number, data: Partial<Pick<SystemDepartment, 'name' | 'parent_id' | 'sort_order' | 'is_active'>>) {
    return request.patch<SystemDepartment>(`/system/departments/${departmentId}`, data)
  },
  listMenus(params?: { page?: number; page_size?: number }) {
    return request.get<PageResponse<SystemMenu>>('/system/menus', { params })
  },
  createMenu(data: {
    parent_id?: number | null
    title: string
    path: string
    component?: string | null
    icon?: string | null
    permission_code?: string | null
    sort_order?: number
    is_visible?: boolean
  }) {
    return request.post<SystemMenu>('/system/menus', data)
  },
  updateMenu(menuId: number, data: Partial<SystemMenu>) {
    return request.patch<SystemMenu>(`/system/menus/${menuId}`, data)
  },
  listOperationLogs(params?: { page?: number; page_size?: number }) {
    return request.get<PageResponse<OperationLog>>('/system/logs/operations', { params })
  },
}
